typing/typemod.ml

```diff
@@ -85,33 +87,112 @@ let extract_sig_open env loc mty =
-      ignore (extract_sig_open env lid.loc md.md_type);
-      assert false
+let type_module_fwd : (Env.t -> Parsetree.module_expr ->
+                       Typedtree.module_expr) ref =
+  ref (fun _ _ -> assert false)
+
+let mod_ident_counter = ref 0
+let generated_module_ident = ref []
+
+let push_current_mid (mi, md, env) =
+  generated_module_ident := (mi, md, env) :: !generated_module_ident
+
+let pop_current_mid () =
+  let slot = List.hd !generated_module_ident in
+  generated_module_ident := List.tl !generated_module_ident;
+  slot
+
+let gen_mod_ident () =
+  let n = !mod_ident_counter in
+  incr mod_ident_counter;
+  let ident = Ident.create (Printf.sprintf "M#%d" n) in
+  ident
+
+let open_struct_level = ref 0
+let in_nested_struct () = !open_struct_level <> 0
+let enter_struct () = incr open_struct_level
+let leave_struct () = decr open_struct_level
+
+let type_open_ ?used_slot ?toplevel ovf env loc me =
+  match me.pmod_desc with
+  | Pmod_functor _ | Pmod_unpack _ | Pmod_extension _ ->
+      raise(Error(me.pmod_loc, env, Invalid_open me))
+  | Pmod_ident lid -> begin
+      let path = Typetexp.lookup_module ~load:true env lid.loc lid.txt in
+      match Env.open_signature ~loc ?used_slot ?toplevel ovf path env with
+      | Some env ->
+          let tme =
+            {
+              mod_desc=Tmod_ident (path, lid);
+              mod_loc=lid.loc;
+              mod_type=Mty_ident path;
+              mod_env=env;
+              mod_attributes=me.pmod_attributes
+            } in
+          tme, env
+      | None ->
+          let md = Env.find_module path env in
+          ignore (extract_sig_open env lid.loc md.md_type);
+          assert false
+    end
+  | _ -> begin
+      enter_struct ();
+      let ident = gen_mod_ident () in
+      let tme = !type_module_fwd env me in
+      leave_struct ();
+      begin
+      match tme.mod_type with
+      | Mty_signature _ -> ()
+      | _ -> raise(Error(me.pmod_loc, env, Invalid_open me));
+      end;
+      let md = {
+        md_type = tme.mod_type;
+        md_loc = me.pmod_loc;
+        md_attributes = me.pmod_attributes;
+      } in
+      let newenv = Env.enter_module_declaration ident md env in
+      push_current_mid (ident, md, newenv);
+      let root = Pident ident in
+      match Env.open_signature ~loc ?used_slot ?toplevel ovf root newenv with
+      | None -> assert false
+      | Some opened_env -> tme, opened_env
+    end
+
+let extract_open od =
+  match od.open_expr.mod_desc with
+  | Tmod_ident (_, _) -> None
+  | Tmod_structure _ | Tmod_apply _ | Tmod_constraint _ ->
+      let id, md, env = pop_current_mid () in
+      let tm =
+        Tstr_module {mb_id=id;
+                     mb_name={txt=Ident.name id; loc=Location.none};
+                     mb_expr = od.open_expr;
+                     mb_attributes=od.open_expr.mod_attributes;
+                     mb_loc=od.open_expr.mod_loc} in
+      Some (tm, md, id, env)
+  | _ -> assert false
```

Comment:
Please list remaining constructors explicitly, it helps reasoning about the code.

Reply:
done